File: parser/encounters/titan_x.py
# parser/encounters/titan_x.py
from __future__ import annotations

from parser.types import Fight, Phase


def build_phases_titan_x(fight: Fight) -> list[Phase]:
    """
    Total = Combat Begin -> Kill Titan X
    BossOnly = first event impliquant Titan X -> Kill Titan X
    """
    boss_name = "Titan X"

    boss_start = None
    for ev in fight.events:
        # un event "impliquant" Titan X (src ou dst) suffit à marquer le pop/engage
        if boss_name.lower() in ev.src.lower() or boss_name.lower() in ev.dst.lower():
            boss_start = ev.ts_sec
            break

    if boss_start is None:
        boss_start = fight.start_sec

    # On retourne 1 phase "Boss" (visible), mais total se calcule via fight.start_sec
    return [
        Phase(
            name="Boss",
            start_sec=boss_start,
            end_sec=fight.end_sec,
            boss_name=boss_name,
        )
    ]

File: rust/src/bosses/titan_x.rs
// src/bosses/titan_x.rs

use crate::types::{CharKind, Fight, Phase};

/// Titan X : phase "Adds" (avant le boss) puis phase "Boss" (Titan X lui-même)
/// La séparation se fait au premier event impliquant "Titan X" directement
pub fn build_phases(fight: &Fight) -> Vec<Phase> {
    // Cherche le premier event où Titan X est attaqué (src = Player, dst contient "Titan X")
    let boss_start = fight.events.iter().find(|ev| {
        ev.src_kind == CharKind::Player
            && ev.dst.to_lowercase().contains("titan x")
            && ev.is_damage()
    });

    match boss_start {
        Some(ev) => {
            let mid = ev.ts_sec;
            let mut phases = Vec::new();

            if mid > fight.start_sec {
                phases.push(Phase {
                    name      : "Adds".to_string(),
                    start_sec : fight.start_sec,
                    end_sec   : mid,
                    boss_name : None, // pas de filtre boss pour la phase adds
                });
            }

            phases.push(Phase {
                name      : "Boss".to_string(),
                start_sec : mid,
                end_sec   : fight.end_sec,
                boss_name : Some("Titan X".to_string()),
            });

            phases
        }
        None => vec![Phase {
            name      : "Boss".to_string(),
            start_sec : fight.start_sec,
            end_sec   : fight.end_sec,
            boss_name : Some("Titan X".to_string()),
        }],
    }
}
